dhis-2/dhis-services/dhis-service-administration/src/main/java/org/hisp/dhis/merge/dataelement/handler/DataDataElementMergeHandler.java

```diff
@@ -0,0 +1,228 @@
+    log.info(
+        "Handling "
+            + sourceDataValuesDuplicates.size()
+            + " duplicate data values, keeping later lastUpdated value");
+
+    // group Data values by key so we can deal with each duplicate correctly
+    Map<String, List<DataValue>> sourceDataValuesGroupedByKey =
+        sourceDataValuesDuplicates.stream()
+            .collect(Collectors.groupingBy(DataDataElementMergeHandler::getDataValueKey));
+
+    // filter groups down to single DV with latest date
+    List<DataValue> filtered =
+        sourceDataValuesGroupedByKey.values().stream()
+            .map(ls -> Collections.max(ls, Comparator.comparing(DataValue::getLastUpdated)))
+            .toList();
+
+    for (DataValue source : filtered) {
+      DataValue matchingTargetDataValue = targetDataValueMap.get(getDataValueKey(source));
+
+      if (matchingTargetDataValue.getLastUpdated().before(source.getLastUpdated())) {
+        dataValueStore.deleteDataValue(matchingTargetDataValue);
+
+        // detaching is required here as it's not possible to add a new DataValue with essentially
+        // the same composite primary key - Throws `NonUniqueObjectException: A different object
+        // with the same identifier value was already associated with the session`
+        entityManager.detach(matchingTargetDataValue);
+        DataValue copyWithNewDataElementRef = DataValue.dataValueWithNewDataElement(source, target);
+        dataValueStore.addDataValue(copyWithNewDataElementRef);
+      }
+    }
+
+    // delete the rest of the source data values after handling the last update duplicate
+    sources.forEach(dataValueStore::deleteDataValues);
+  }
+
+  /**
+   * Method to handle merging non-duplicate {@link DataValue}s. A new {@link DataValue} will be
+   * created from the old {@link DataValue} values, and it will use the target {@link DataElement}
+   * ref. This new {@link DataValue} will be saved to the database. This sequence is required as a
+   * {@link DataValue} has a composite primary key which includes {@link DataElement}. This
+   * prohibits updating the {@link DataElement} ref in a source {@link DataValue}.
+   *
+   * <p>All source {@link DataValue}s will then be deleted.
+   *
+   * @param dataValues {@link DataValue}s to merge
+   * @param sources source {@link DataElement}s
+   * @param target target {@link DataElement}
+   */
+  private void handleNonDuplicates(
+      @Nonnull List<DataValue> dataValues,
+      @Nonnull List<DataElement> sources,
+      @Nonnull DataElement target) {
+    log.info(
+        "Handling "
+            + dataValues.size()
+            + " non duplicate data values. Add new DataValue entry (using target DataElement ref) and delete old source entry");
+
+    dataValues.forEach(
+        sourceDataValue -> {
+          DataValue copyWithNewDataElementRef =
+              DataValue.dataValueWithNewDataElement(sourceDataValue, target);
+
+          dataValueStore.addDataValue(copyWithNewDataElementRef);
```

Comment:
- Do we have an estimation of how many dataValues would be saved in this for loop ?
- If the number is too big, I think we could consider to have implement batch insert here.

Reply:
yep, I thought about batching here. After seeing the basic performance tests results I decided not to go down that road. Tests were done with 155k duplicate `DataValue`s and the merge was finishing in ~10 seconds (details in PR description)
This might be a `let's see what feedback we get on this` situation and only implement batching if needed. Definitely worth mentioning this though.